README.md
shapecode-php
=============
*A Shapeways API library in PHP.*

Copyright (C) 2014-2016 Jublo Solutions <support@jublo.net>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

### Requirements

- PHP 5.5.0 or higher
- JSON extension
- OpenSSL extension


Authentication
--------------

To authenticate your API requests on behalf of a certain Shapeways user
(following OAuth 1.0a), take a look at these steps:

```php
require_once ('shapecode.php');
Shapecode::setConsumerKey('YOURKEY', 'YOURSECRET'); // static, see 'Using multiple Shapecode instances'

$sc = Shapecode::getInstance();
```

You may either set the OAuth token and secret, if you already have them:
```php
$sc->setToken('YOURTOKEN', 'YOURTOKENSECRET');
```

Or you authenticate, like this:

```php
session_start();

if (! isset($_SESSION['oauth_token'])) {
    // get the request token
    $reply = $sc->oauth1_requestToken(array(
        'oauth_callback' => 'http://' . $_SERVER['HTTP_HOST'] . $_SERVER['REQUEST_URI']
    ));

    // store the token
    $_SESSION['oauth_token'] = $reply->oauth_token;
    $_SESSION['oauth_token_secret'] = $reply->oauth_token_secret;
    $_SESSION['oauth_verify'] = true;

    // redirect to auth website
    header('Location: ' . $reply->authentication_url);
    die();

} elseif (isset($_GET['oauth_verifier']) && isset($_SESSION['oauth_verify'])) {
    // verify the token
    $sc->setToken($_SESSION['oauth_token'], $_SESSION['oauth_token_secret']);
    unset($_SESSION['oauth_verify']);

    // get the access token
    $reply = $sc->oauth1_accessToken(array(
        'oauth_verifier' => $_GET['oauth_verifier']
    ));

    // store the token (which is different from the request token!)
    $_SESSION['oauth_token'] = $reply->oauth_token;
    $_SESSION['oauth_token_secret'] = $reply->oauth_token_secret;

    // send to same URL, without oauth GET parameters
    header('Location: ' . basename(__FILE__));
    die();
}

// assign access token on each page load
$sc->setToken($_SESSION['oauth_token'], $_SESSION['oauth_token_secret']);
```

Usage examples
--------------

When you have an access token, calling the API is simple:

```php
$sc->setToken($_SESSION['oauth_token'], $_SESSION['oauth_token_secret']); // see above

$reply = (array) $sc->api();
print_r($reply);
```

Adding a model to your cart is as easy as this:

```php
$reply = $sc->orders_cart('modelId=480903');
```

For more complex parameters (see the [Shapeways API documentation](https://developers.shapeways.com/)),
giving all parameters in an array is supported, too:

```php
$params = array(
    'modelId' => '480903',
    'materialId' => 61,
    'quantity' => 3
);
$reply = $sc->orders_cart($params);
```

When **uploading files to Shapeways**, just give the file path:

```php
$params = array(
    'file' => 'in-some-folder/there-is/the-model.stl',
    'fileName' => 'the-model.stl',
    'hasRightsToModel' => 1,
    'acceptTermsAndConditions' => 1,
    'title' => 'My great model',
    'description' => 'Lorem ipsum dolor sit amet',
    'isPublic' => 1,
    'isForSale' => 1,
    'isDownloadable' => 0,
    'tags' => array(
        'ideas',
        'miniatures',
        'stuff'
    )
);
$reply = $sc->models($params); // required HTTP POST is auto-detected
```

Mapping API methods to Shapecode function calls
-----------------------------------------------

As you can see from the last example, there is a general way how the Shapeways
API methods map to Shapecode function calls. The general rules are:

1. Omit the version info in the function name.

    Example: ```/v1``` is not part of the Shapecode function call.

2. For each slash in a Shapeways API method, use an underscore in the Shapecode function.

    Example: ```orders/cart/v1``` maps to ```Shapecode::orders_cart()```.

3. For each underscore in a Shapeways API method, use camelCase in the Shapecode function.

    Example: ```oauth1/request_token/v1``` maps to ```Shapecode::oauth1_requestToken()```.

4. For each parameter template in method, use UPPERCASE in the Shapecode function.
    Also don’t forget to include the parameter in your parameter list.

    Example:
    - ```materials/{materialId}/v1``` maps to ```Shapecode::materials_MATERIALID('materialId=73')```.

HTTP methods (GET, POST, DELETE etc.)
-------------------------------------

Never care about which HTTP method (verb) to use when calling a Shapeways API.
Shapecode is intelligent enough to find out on its own.
For the automatic detection to work, be sure to use the correct required parameters,
as outlined in the [Shapeways API documentation](https://developers.shapeways.com/).

The only exception to the above is the `DELETE models/{modelId}` method.
To call it, use the `delete=1` parameter. It will trigger the DELETE method,
but is not sent to the API.

Response codes
--------------

The HTTP response code that the API gave is included in any return values.
You can find it within the return object’s ```httpstatus``` property.

To know whether your API call was successful, check the ```$reply->result``` string,
which either reads ```success``` or ```failure```.

Return formats
--------------
The default return format for API calls is a PHP object.
Upon your choice, you may also get PHP arrays directly:

```php
$sc->setReturnFormat(SHAPECODE_RETURNFORMAT_ARRAY);
```

The Shapeways API natively responds to API calls in JSON (JS Object Notation).
To get a JSON string, set the corresponding return format:

```php
$sc->setReturnFormat(SHAPECODE_RETURNFORMAT_JSON);
```

Using multiple Shapecode instances
----------------------------------

By default, Shapecode works with just one instance. This programming paradigma is
called a *singleton*.

Getting the main Shapecode object is done like this:

```php
$sc = Shapecode::getInstance();
```

If you need to run requests to the Shapeways API for multiple users at once,
Shapecode supports this as well. Instead of getting the instance like shown above,
create a new object:

```php
$sc1 = new Shapecode;
$sc2 = new Shapecode;
```

Please note that your OAuth consumer key and secret is shared within
multiple Shapecode instances, while the OAuth request and access tokens with their
secrets are *not* shared.

How Do I…?
----------

### …walk through paged results?

The Shapeways API utilizes a technique called ‘paging’ for
large result sets. Pages separates results into pages of no more than
36 results at a time, and provides a means to move backwards and
forwards through these pages.

Here is how you can walk through paged results with Shapecode.

1. Get the first result set of a paged method:
```php
$page = 1;
$result1 = $sc->models();
```

2. To navigate forth, increment the ```$page```:
```php
$page++;
```

3. If ```$nextCursor``` is not 0, use this cursor to request the next result page:
```php
    $result2 = $sc->models("page=$page");
```

It might make sense to use the pages in a loop.  Watch out, though,
not to send more than the allowed number of requests
per rate-limit timeframe, or else you will hit your rate-limit.

### …know what cacert.pem is for?

Connections to the Shapeways API are done over a secured SSL connection.
Shapecode-php checks if the Shapeways API server has a valid SSL certificate.
Valid certificates have a correct signature-chain.
The cacert.pem file contains a list of all public certificates for root
certificate authorities. You can find more information about this file
at http://curl.haxx.se/docs/caextract.html.

### …set the timeout for requests to the Shapeways API?

For connecting to Shapeways, Shapecode uses the cURL library, if available.
You can specify both the connection timeout and the request timeout,
in milliseconds:

```php
$sc->setConnectionTimeout(2000);
$sc->setTimeout(5000);
```

If you don't specify the timeout, Shapecode uses these values:

- connection time = 5000 ms = 5 s
- timeout = 2000 ms = 2 s

### …disable cURL?

Shapecode automatically detects whether you have the PHP cURL extension enabled.
If not, the library will try to connect to the Shapeways API via socket.
For this to work, the PHP setting `allow_url_fopen` must be enabled.

You may also manually disable cURL.  Use the following call:

```php
$sc->setUseCurl(false);
```


CHANGELOG
shapecode-php - changelog
=========================

1.1.2 (2016-08-10)
+ Add POST orders/{orderId}/reprint method

1.1.1 (2015-12-06)
+ Fix multiple inline parameter detection

1.1.0 (2015-12-04)
+ Better Proxy HTTP status detection
+ Don't require cURL, allow stream connections too
+ Use default timeout
+ Add composer/installers to allow install path changes
+ Update materials list
+ #13 Orders API
+ #12 Allow for multiple parameters in templated methods
  by replacing preg_match with preg_match_all
+ Update cacert.pem

1.0.0 (2014-02-27)
Initial release


examples/custom_function.php
<?php

/**
 * Uploads a 3D model
 *
 * @param string $file                 Path to the model to upload
 * @param string $name                 Model name
 * @param string $description          Model description
 * @param string $tags                 Comma-separated list of tags to apply
 * @param string $default_material     Name of the default material
 * @param array  $disallowed_materials Array of materials not allowed
 * @param float  $markup               Price markup to earn
 * @param string $consumer_key         OAuth app key
 * @param string $consumer_secret      OAuth app secret
 * @param string $access_token         OAuth access token to the account to upload to
 * @param string $access_token_secret  OAuth access token secret
 *
 * @return array API data
 */
function upload3DFile(
    $file,
    $name,
    $description,
    $tags,
    $default_material,
    $disallowed_materials,
    $markup,
    $consumer_key,
    $consumer_secret,
    $access_token,
    $access_token_secret
)
{
    // materials as of 2015-03-08
    $all_materials = (array) json_decode('
        {
            "White Strong & Flexible": 6,
            "White Strong & Flexible Polished": 62,
            "Green Strong & Flexible Polished": 94,
            "Orange Strong & Flexible Polished": 95,
            "Yellow Strong & Flexible Polished": 93,
            "Black Strong & Flexible": 25,
            "Red Strong & Flexible Polished": 76,
            "Pink Strong & Flexible Polished": 77,
            "Blue Strong & Flexible Polished": 78,
            "Purple Strong & Flexible Polished": 75,
            "Elasto Plastic": 82,
            "Frosted Ultra Detail": 61,
            "Frosted Detail": 60,
            "White Detail": 5,
            "Black Detail": 7,
            "Transparent Detail": 4,
            "Full Color Sandstone": 26,
            "Sandstone": 27,
            "Raw Bronze": 86,
            "14K Gold": 91,
            "14k Gold Plated": 110,
            "Rhodium Plated": 113,
            "Gloss Cobalt Blue Porcelain": 106,
            "Matte Black Steel": 89,
            "18k Gold Plated": 112,
            "Polished Bronze": 87,
            "Raw Brass": 84,
            "Gloss Celadon Green Porcelain": 104,
            "Polished Brass": 85,
            "Polished Nickel Steel": 88,
            "14k Rose Gold Plated": 111,
            "Polished Grey Steel": 90,
            "Gloss Oribe Green Porcelain": 105,
            "Gloss Yellow Porcelain": 108,
            "14k Rose Gold": 96,
            "Gloss Red Porcelain": 109,
            "Gloss Blue Porcelain": 107,
            "14k White Gold": 97,
            "Platinum": 99,
            "18k Gold": 98,
            "Full Color Plastic": 100,
            "Matte Black Porcelain": 101,
            "Gloss Black Porcelain": 102,
            "Gloss White Porcelain": 103,
            "Premium Silver": 81,
            "Polished Silver": 54,
            "Raw Silver": 53,
            "Stainless Steel": 23,
            "Matte Gold Steel": 31,
            "Polished Gold Steel": 39,
            "Matte Bronze Steel": 37,
            "Polished Bronze Steel": 38,
            "Metallic Plastic": 28,
            "Polished Metallic Plastic": 66,
            "Castable Wax": 92
        }
    ');
    $all_material_names = array_keys($all_materials);

    require_once('../src/shapecode.php');
    Shapecode::setConsumerKey($consumer_key, $consumer_secret);
    $sc = Shapecode::getInstance();
    $sc->setToken($access_token, $access_token_secret);

    // comma-separated list of tags?
    if (is_string($tags)) {
        $tags = explode(',', $tags);
        // remove leading and trailing spaces
        array_walk($tags, function(&$val){$val = trim($val);});
    }

    // validate default material
    if (! in_array($default_material, $all_material_names)) {
        throw new Exception('Unknown default material: "' . htmlspecialchars($default_material) . '"');
        return false;
    }

    // validate markup
    if (! is_numeric($markup)) {
        throw new Exception('Non-numeric markup: "' . htmlspecialchars($markup) . '"');
        return false;
    }

    // build materials list
    $materials = array();
    foreach ($all_materials as $material_name => $material_id) {

        // skip disallowed materials
        if (is_array($disallowed_materials)
            && in_array($material_name, $disallowed_materials)
        ) {
            continue;
        }

        $materials[$material_id] = array(
            'id' => $material_id,
            'markup' => $markup,
            'isActive' => 1
        );
    }

    return $sc->models(array(
        'file' => $file,
        'fileName' => basename($file),
        'hasRightsToModel' => 1,
        'acceptTermsAndConditions' => 1,
        'title' => $name,
        'description' => $description,
        'isPublic' => 0,
        'isForSale' => 0,
        'isDownloadable' => 0,
        'tags' => $tags,
        'materials' => $materials,
        'defaultMaterialId' => $all_materials[$default_material] // use ID
    ));
}

// get OAuth tokens from external file, not included in repo
require_once('private_shapeways_tokens.php');

$upload_data = upload3DFile(
    '../demo-data/cube-1cm3-centered_in_meter.stl', // path to the file to upload
    'Test Model', // name of the 3D model
    'This is a test model showing a cube.', // description of the 3D model
    'test, demo, coding, cube', // provided like « a,b,c »...
    'Royal Blue Strong & Flexible Polished', // defaultMaterialName
    array('Avocado Green Ceramics', 'Raw Bronze'), // $excludedMaterials, array with all material names not allowed
    1.99, // $markup, same markup for each material

    // for Oauth protocol (correct me if this is wrong)
    SHAPEWAYS_CONSUMER_KEY, // $consumer_key,
    SHAPEWAYS_CONSUMER_SECRET, // $consumer_secret
    SHAPEWAYS_ACCESS_TOKEN, // $access_token,
    SHAPEWAYS_ACCESS_TOKEN_SECRET // $access_token_secret
);

// here you can check if the upload worked
if ($upload_data->result === 'success') {
    // maybe store the modelId in your database?
    echo $upload_data->modelId;
}


examples/shapecode-demo/index.php
<?php

/**
 * Uploads a 3D model
 *
 * @param string $file                 Path to the model to upload
 * @param string $name                 Model name
 * @param string $description          Model description
 * @param string $tags                 Comma-separated list of tags to apply
 * @param string $default_material     Name of the default material
 * @param array  $disallowed_materials Array of materials not allowed
 * @param float  $markup               Price markup to earn
 * @param string $consumer_key         OAuth app key
 * @param string $consumer_secret      OAuth app secret
 * @param string $access_token         OAuth access token to the account to upload to
 * @param string $access_token_secret  OAuth access token secret
 *
 * @return array API data
 */
function upload3DFile(
    $file,
    $filename,
    $name,
    $description,
    $tags,
    $default_material,
    $disallowed_materials,
    $markup,
    $consumer_key,
    $consumer_secret,
    $access_token,
    $access_token_secret
)
{
    // materials as of 2014-06-24
    $all_materials = (array) json_decode('
        {
            "White Strong & Flexible": 6,
            "White Strong & Flexible Polished": 62,
            "Orange Strong & Flexible Polished": 95,
            "Green Strong & Flexible Polished": 94,
            "Yellow Strong & Flexible Polished": 93,
            "Black Strong & Flexible": 25,
            "Red Strong & Flexible Polished": 76,
            "Pink Strong & Flexible Polished": 77,
            "Blue Strong & Flexible Polished": 78,
            "Purple Strong & Flexible Polished": 75,
            "Elasto Plastic": 82,
            "Frosted Ultra Detail": 61,
            "Frosted Detail": 60,
            "White Detail": 5,
            "Black Detail": 7,
            "Transparent Detail": 4,
            "Full Color Sandstone": 26,
            "Sandstone": 27,
            "Glazed Ceramics": 63,
            "Gloss Black Ceramics": 64,
            "Satin Black Ceramics": 70,
            "Pastel Yellow Ceramics": 74,
            "Eggshell Blue Ceramics": 72,
            "Avocado Green Ceramics": 73,
            "Raw Brass": 84,
            "Raw Bronze": 86,
            "Polished Brass": 85,
            "14K Gold": 91,
            "Gold Plated Brass": 83,
            "Polished Grey Steel": 90,
            "Matte Black Steel": 89,
            "Polished Bronze": 87,
            "Polished Nickel Steel": 88,
            "Premium Silver": 81,
            "Polished Silver": 54,
            "Raw Silver": 53,
            "Stainless Steel": 23,
            "Matte Gold Steel": 31,
            "Polished Gold Steel": 39,
            "Matte Bronze Steel": 37,
            "Polished Bronze Steel": 38,
            "Metallic Plastic": 28,
            "Polished Metallic Plastic": 66,
            "Castable Wax": 92
        }
    ');
    $all_material_names = array_keys($all_materials);

    $sc = Shapecode::getInstance();
    $sc->setToken($access_token, $access_token_secret);

    // comma-separated list of tags?
    if (is_string($tags)) {
        $tags = explode(',', $tags);
        // remove leading and trailing spaces
        for ($i = 0; $i < count($tags); $i++) {
            $tags[$i] = trim($tags[$i]);
        }
    }

    // validate default material
    if (! in_array($default_material, $all_material_names)) {
        throw new Exception('Unknown default material: "' . htmlspecialchars($default_material) . '"');
        return false;
    }

    // validate markup
    if (! is_numeric($markup)) {
        throw new Exception('Non-numeric markup: "' . htmlspecialchars($markup) . '"');
        return false;
    }

    // build materials list
    $materials = array();
    foreach ($all_materials as $material_name => $material_id) {

        // skip disallowed materials
        if (is_array($disallowed_materials)
            && in_array($material_name, $disallowed_materials)
        ) {
            continue;
        }

        $materials[$material_id] = array(
            'id' => $material_id,
            'markup' => $markup,
            'isActive' => 1
        );
    }

    return $sc->models(array(
        'file' => $file,
        'fileName' => $filename,
        'hasRightsToModel' => 1,
        'acceptTermsAndConditions' => 1,
        'title' => $name,
        'description' => $description,
        'isPublic' => 0,
        'isForSale' => 0,
        'isDownloadable' => 0,
        'tags' => $tags,
        'materials' => $materials,
        'defaultMaterialId' => $all_materials[$default_material] // use ID
    ));
}

error_reporting(E_ALL);
ini_set('display_errors', 1);

session_start();

if (isset($_REQUEST['reset'])) {
    session_destroy();
}

if (! isset($_SESSION['oauth_token']) && ! isset($_SESSION['oauth_verify']) && ! isset($_POST['authenticate'])) {
    ?><!DOCTYPE HTML>
<html>
    <head>
        <title>Shapeways upload demo</title>
    </head>
    <body>
        <p>You are not logged in.</p>
        <form method="post">
            <button type="submit" name="authenticate" value="1">Login</button>
        </form>
    </body>
</html>
    <?php
    die();
}

// get OAuth tokens from external file, not included in repo
require_once('private_shapeways_tokens.php');

if (isset($_FILES['upload']) && $_FILES['upload']['error'] === 0) {

    $upload_data = upload3DFile(
        $_FILES['upload']['tmp_name'], // path to the file to upload
        $_FILES['upload']['name'],
        $_POST['title'], // name of the 3D model
        isset($_POST['description']) ? $_POST['description'] : '', // description of the 3D model
        isset($_POST['tags']) ? $_POST['tags'] : '', // provided like « a,b,c »...
        'Black Strong & Flexible', // defaultMaterialName
        array('Avocado Green Ceramics', 'Raw Bronze'), // $excludedMaterials, array with all material names not allowed
        isset($_POST['markup']) ? $_POST['markup'] : '', // $markup, same markup for each material

        // for Oauth protocol (correct me if this is wrong)
        SHAPEWAYS_CONSUMER_KEY, // $consumer_key,
        SHAPEWAYS_CONSUMER_SECRET, // $consumer_secret
        SHAPEWAYS_ACCESS_TOKEN, // $access_token,
        SHAPEWAYS_ACCESS_TOKEN_SECRET // $access_token_secret
    );

    if ($upload_data->result === 'success') {
        ?><!DOCTYPE HTML>
<html>
    <head>
        <title>Shapeways upload demo</title>
    </head>
    <body>
        <p>You are logged in.</p>
        <form method="post">
            <button type="submit" name="reset" value="1">Logout</button>
        </form>
        <hr />
        <p>Model uploaded successfully.</p>
        <p><a href="http://www.shapeways.com/model/<?php echo $upload_data->modelId; ?>/?key=<?php echo $upload_data->secretKey; ?>" target="_blank">Take a look</a></p>
    </body>
</html>
        <?php
    } elseif ($upload_data->result === 'failure') {
        ?><!DOCTYPE HTML>
<html>
    <head>
        <title>Shapeways upload demo</title>
    </head>
    <body>
        <p>You are logged in.</p>
        <form method="post">
            <button type="submit" name="reset" value="1">Logout</button>
        </form>
        <hr />
        <p>Error in model upload.</p>
        <p><pre><?php print_r($upload_data); ?></pre></p>
    </body>
</html>
        <?php
    }
    die();
}

?><!DOCTYPE HTML>
<html>
    <head>
        <title>Shapeways upload demo</title>
    </head>
    <body>
        <p>You are logged in.</p>
        <form method="post">
            <button type="submit" name="reset" value="1">Logout</button>
        </form>
        <hr />
        <form method="post" enctype="multipart/form-data">
            <label for="upload">Select model file to upload:</label>
            <input type="file" name="upload" id="upload" required /><br />
            <label for="title">Model title:</label>
            <input type="text" name="title" id="title" required /><br />
            <label for="description">Model description (optional)</label>
            <input type="text" name="description" id="description" /><br />
            <label for="tags">Model tags (optional)</label>
            <input type="text" name="tags" id="tags" /><br />
            <label for="markup">Material markup</label>
            <input type="text" name="markup" id="markup" required value="0.00" /><br />
            <button type="submit">Upload</button>
        </form>
    </body>
</html>
